File: wpimath/src/test/java/edu/wpi/first/math/kinematics/MecanumDriveWheelSpeedsSerdeTest.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package edu.wpi.first.math.kinematics;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;

import edu.wpi.first.math.proto..ProtobufMecanumDriveWheelSpeeds;
import edu.wpi.first.util.struct.Struct;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import org.junit.jupiter.api.Test;
import java.util.Arrays;

public class MecanumDriveWheelSpeedsSerdeTest {
  private static final MecanumDriveWheelSpeeds DATA = new MecanumDriveWheelSpeeds(1.91, 2.29);
  private static final byte[] STRUCT_BUFFER = new byte[]{63, -2, -113, 92, 40, -11, -62, -113, 64, 2, 81, -21, -123, 30, -72, 82};

  @Test
  void testStructPack() {
    ByteBuffer buffer = ByteBuffer.allocate(MecanumDriveWheelSpeeds.struct.getSize());
    buffer.order(ByteOrder.LITTLE_ENDIAN);
    MecanumDriveWheelSpeeds.struct.pack(buffer, DATA);

    byte[] actual = buffer.array();
    String newContent = new String(buffer.array());
    System.out.println(Arrays.toString(actual));
    System.out.println(newContent);
    assertArrayEquals(actual, STRUCT_BUFFER);
  }

  @Test
  void testStructUnpack() {
    ByteBuffer buffer = ByteBuffer.wrap(STRUCT_BUFFER);
    buffer.order(ByteOrder.LITTLE_ENDIAN);

    MecanumDriveWheelSpeeds data = MecanumDriveWheelSpeeds.struct.unpack(buffer);
    assertEquals(DATA.frontLeftMetersPerSecond, data.frontLeftMetersPerSecond);
    assertEquals(DATA.frontRightMetersPerSecond, data.frontRightMetersPerSecond);
    assertEquals(DATA.rearLeftMetersPerSecond, data.rearLeftMetersPerSecond);
    assertEquals(DATA.rearRightMetersPerSecond, data.rearRightMetersPerSecond);
  }

  @Test
  void testProtoPack() {
    ProtobufMecanumDriveWheelSpeeds proto = MecanumDriveWheelSpeeds.proto.createMessage();
    MecanumDriveWheelSpeeds.proto.pack(proto, DATA);

    assertEquals(DATA.frontLeftMetersPerSecond, proto.getFrontLeftMps());
    assertEquals(DATA.frontRightMetersPerSecond, proto.getFrontRightMps());
    assertEquals(DATA.rearLeftMetersPerSecond, proto.getRearLeftMps());
    assertEquals(DATA.rearRightMetersPerSecond, proto.getRearRightMps());
  }

  @Test
  void testProtoUnpack() {
    ProtobufMecanumDriveWheelSpeeds proto = MecanumDriveWheelSpeeds.proto.createMessage();
    proto.setFrontLeftMps(DATA.frontLeftMetersPerSecond);
    proto.setFrontRightMps(DATA.frontRightMetersPerSecond);
    proto.setRearLeftMps(DATA.rearLeftMetersPerSecond);
    proto.setRearRightMps(DATA.rearRightMetersPerSecond);

    MecanumDriveWheelSpeeds data = MecanumDriveWheelSpeeds.proto.unpack(proto);
    assertEquals(DATA.frontLeftMetersPerSecond, data.frontLeftMetersPerSecond);
    assertEquals(DATA.frontRightMetersPerSecond, data.frontRightMetersPerSecond);
    assertEquals(DATA.rearLeftMetersPerSecond, data.rearLeftMetersPerSecond);
    assertEquals(DATA.rearRightMetersPerSecond, data.rearRightMetersPerSecond);
  }
}

File: wpimath/src/test/native/cpp/kinematics/serde/MecanumDriveWheelSpeedsSerdeTest.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include <gtest/gtest.h>

#include "frc/geometry/MecanumDriveWheelSpeeds.h"
#include "geometry2d.pb.h"

using namespace frc;

namespace {

using StructType = wpi::Struct<frc::MecanumDriveWheelSpeeds>;
using ProtoType = wpi::Protobuf<frc::MecanumDriveWheelSpeeds>;

constexpr std::array<uint8_t, StructType::kSize> create_test_buffer() {
  std::array<uint8_t, StructType::kSize> output;
  int buffer[] = {
    0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0,
};
  for (size_t idx = 0; idx < StructType::kSize; ++idx) {
    output[idx] = static_cast<uint8_t>(buffer[idx]);
  }
  return output;
}

std::array<uint8_t, StructType::kSize> kExpectedStructBuffer =
    create_test_buffer();

constexpr MecanumDriveWheelSpeeds kExpectedData{};
}  // namespace

TEST(MecanumDriveWheelSpeedsTest, StructPack) {
  uint8_t buffer[StructType::kSize];
  StructType::Pack(buffer, kExpectedData);

  for (size_t i = 0; i < StructType::kSize; ++i) {
    EXPECT_EQ(kExpectedStructBuffer[i], buffer[i]) << " on byte " << i;
  }
}

TEST(MecanumDriveWheelSpeedsTest, StructUnpack) {
  MecanumDriveWheelSpeeds unpacked_data = StructType::Unpack(kExpectedStructBuffer);

    EXPECT_EQ(kExpectedData.FrontLeftMps(), unpacked_data.FrontLeftMps());
    EXPECT_EQ(kExpectedData.FrontRightMps(), unpacked_data.FrontRightMps());
    EXPECT_EQ(kExpectedData.RearLeftMps(), unpacked_data.RearLeftMps());
    EXPECT_EQ(kExpectedData.RearRightMps(), unpacked_data.RearRightMps());
}


TEST(MecanumDriveWheelSpeedsTest, ProtobufPack) {
  wpi::proto::ProtobufMecanumDriveWheelSpeeds proto;
  ProtoType::Pack(&proto, kExpectedData);

    EXPECT_EQ(kExpectedData.FrontLeftMps(), proto.front_left_mps());
    EXPECT_EQ(kExpectedData.FrontRightMps(), proto.front_right_mps());
    EXPECT_EQ(kExpectedData.RearLeftMps(), proto.rear_left_mps());
    EXPECT_EQ(kExpectedData.RearRightMps(), proto.rear_right_mps());
}



TEST(MecanumDriveWheelSpeedsTest, ProtobufUnpack) {
  wpi::proto::ProtobufMecanumDriveWheelSpeeds proto;
  proto.set_front_left_mps(kExpectedData.FrontLeftMps());
  proto.set_front_right_mps(kExpectedData.FrontRightMps());
  proto.set_rear_left_mps(kExpectedData.RearLeftMps());
  proto.set_rear_right_mps(kExpectedData.RearRightMps());

  MecanumDriveWheelSpeeds unpacked_data = ProtoType::Unpack(proto);
    EXPECT_EQ(kExpectedData.FrontLeftMps(), unpacked_data.FrontLeftMps());
    EXPECT_EQ(kExpectedData.FrontRightMps(), unpacked_data.FrontRightMps());
    EXPECT_EQ(kExpectedData.RearLeftMps(), unpacked_data.RearLeftMps());
    EXPECT_EQ(kExpectedData.RearRightMps(), unpacked_data.RearRightMps());
}